File: windows/app/OpenBurnBar.App.Community/CommunityCityKey.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

namespace OpenBurnBar.App.Community;

/// <summary>
/// Ports <c>functions/src/community/geo.ts</c> city key canonicalization (asciiFold, slugifyCity, canonicalizeCityKey).
/// </summary>
public static class CommunityCityKey
{
    private static readonly Regex CombiningMarks = new(@"[\u0300-\u036f]+", RegexOptions.Compiled);
    private static readonly Regex NonSlug = new(@"[^a-z0-9]+", RegexOptions.Compiled);
    private static readonly Regex EdgeDashes = new(@"^-+|-+$", RegexOptions.Compiled);

    /// <summary>Characters NFD does not reduce to ASCII — replaced before FormD (ports geo.ts).</summary>
    private static readonly Dictionary<char, string> NonDecomposable = new()
    {
        ['\u00D8'] = "O", ['\u00F8'] = "o",
        ['\u0141'] = "L", ['\u0142'] = "l",
        ['\u00D0'] = "D", ['\u00F0'] = "d",
        ['\u00DE'] = "T", ['\u00FE'] = "t",
        ['\u00DF'] = "ss",
        ['\u0130'] = "I", ['\u0131'] = "i",
        ['\u0110'] = "D", ['\u0111'] = "d",
        ['\u014A'] = "N", ['\u014B'] = "n",
        ['\u017D'] = "Z", ['\u017E'] = "z",
        ['\u0160'] = "S", ['\u0161'] = "s",
        ['\u015A'] = "S", ['\u015B'] = "s",
        ['\u017B'] = "Z", ['\u017C'] = "z",
        ['\u0106'] = "C", ['\u0107'] = "c",
        ['\u010C'] = "C", ['\u010D'] = "c",
        ['\u0158'] = "R", ['\u0159'] = "r",
        ['\u016E'] = "U", ['\u016F'] = "u",
        ['\u0147'] = "N", ['\u0148'] = "n",
        ['\u010E'] = "D", ['\u010F'] = "d",
        ['\u0164'] = "T", ['\u0165'] = "t",
    };

    public static string AsciiFold(string input)
    {
        var replaced = new StringBuilder(input.Length);
        foreach (var ch in input)
        {
            if (NonDecomposable.TryGetValue(ch, out var mapped))
            {
                replaced.Append(mapped);
            }
            else
            {
                replaced.Append(ch);
            }
        }

        return CombiningMarks.Replace(replaced.ToString().Normalize(NormalizationForm.FormD), "");
    }

    public static string SlugifyCity(string cityName)
    {
        var folded = AsciiFold(cityName).ToLowerInvariant();
        var slug = NonSlug.Replace(folded, "-");
        slug = EdgeDashes.Replace(slug, "");
        if (slug.Length > 40)
        {
            slug = slug[..40].TrimEnd('-');
        }

        return slug;
    }

    public static string CanonicalizeCityKey(string cityName, string countryCode, string regionCode)
    {
        var cc = countryCode.Trim().ToUpperInvariant();
        var rc = regionCode.Trim().ToUpperInvariant();
        if (rc.StartsWith($"{cc}-", StringComparison.Ordinal))
        {
            rc = rc[(cc.Length + 1)..];
        }

        return $"{cc}-{rc}-{SlugifyCity(cityName)}";
    }
}

File: android/app/src/main/java/com/openburnbar/data/community/CommunityCityKey.kt
package com.openburnbar.data.community

import java.text.Normalizer
import java.util.Locale
import java.util.regex.Pattern

/**
 * Ports [functions/src/community/geo.ts] city key canonicalization
 * (`asciiFold`, `slugifyCity`, `canonicalizeCityKey`).
 */
object CommunityCityKey {
    private val NON_SLUG = Pattern.compile("[^a-z0-9]+")
    private val EDGE_DASHES = Pattern.compile("^-+|-+$")

    /** Characters NFD does not reduce to ASCII — replaced before decomposition (ports geo.ts). */
    private val NON_DECOMPOSABLE: Map<Char, String> =
        mapOf(
            '\u00D8' to "O", '\u00F8' to "o",
            '\u0141' to "L", '\u0142' to "l",
            '\u00D0' to "D", '\u00F0' to "d",
            '\u00DE' to "T", '\u00FE' to "t",
            '\u00DF' to "ss",
            '\u0130' to "I", '\u0131' to "i",
            '\u0110' to "D", '\u0111' to "d",
            '\u014A' to "N", '\u014B' to "n",
            '\u017D' to "Z", '\u017E' to "z",
            '\u0160' to "S", '\u0161' to "s",
            '\u015A' to "S", '\u015B' to "s",
            '\u017B' to "Z", '\u017C' to "z",
            '\u0106' to "C", '\u0107' to "c",
            '\u010C' to "C", '\u010D' to "c",
            '\u0158' to "R", '\u0159' to "r",
            '\u016E' to "U", '\u016F' to "u",
            '\u0147' to "N", '\u0148' to "n",
            '\u010E' to "D", '\u010F' to "d",
            '\u0164' to "T", '\u0165' to "t",
        )

    fun asciiFold(input: String): String {
        val replaced = buildString(input.length) {
            for (ch in input) {
                append(NON_DECOMPOSABLE[ch] ?: ch)
            }
        }
        val normalized = Normalizer.normalize(replaced, Normalizer.Form.NFD)
        val sb = StringBuilder(normalized.length)
        for (ch in normalized) {
            if (Character.getType(ch) != Character.NON_SPACING_MARK.toInt()) {
                sb.append(ch)
            }
        }
        return sb.toString()
    }

    fun slugifyCity(cityName: String): String {
        var slug =
            NON_SLUG
                .matcher(asciiFold(cityName).lowercase(Locale.ROOT))
                .replaceAll("-")
        slug = EDGE_DASHES.matcher(slug).replaceAll("")
        if (slug.length > 40) {
            slug = slug.take(40).trimEnd('-')
        }
        return slug
    }

    fun canonicalizeCityKey(cityName: String, countryCode: String, regionCode: String): String {
        val cc = countryCode.trim().uppercase(Locale.ROOT)
        var rc = regionCode.trim().uppercase(Locale.ROOT)
        if (rc.startsWith("$cc-")) {
            rc = rc.substring(cc.length + 1)
        }
        return "$cc-$rc-${slugifyCity(cityName)}"
    }
}
